File: prototypeLeger/app/src/main/java/com/example/scrabbleprototype/model/UserStats.kt
package com.example.scrabbleprototype.model

import android.util.Log

class UserStats {
    var gamesPlayed: Int = 0
    var gamesWon: Int = 0
    var totalPoints: Int = 0
    var totalTimeMS: Int = 0

    var averagePoints: Int = 0
    var averageTime: Int = 0

    var logins = arrayListOf<Connection>()
    var logouts = arrayListOf<Connection>()

    var games = arrayListOf<Game>()

    @JvmName("getUserAveragePoints")
    fun getAveragePoints(): Int {
        return totalPoints / gamesPlayed
    }

    @JvmName("getUserAverageTime")
    fun getAverageTime(): String {
        val averageTimeMS = totalTimeMS / gamesPlayed
        val hour = (averageTimeMS / (1000 * 60 * 60)) % 24
        val minute = (averageTimeMS / (1000 * 60)) % 60
        val second = (averageTimeMS / 1000) % 60
        if(hour == 0) return String.format("%02dm %02ds", minute, second)
        return String.format("%02dh %02dm %02ds", hour, minute, second)
    }
}

File: client/src/app/classes/user-stats.ts
/* eslint-disable max-classes-per-file */
import { Connection, GameDB, UserStatsDB } from '@common/user-stats';
export class Game {
    startDate: string;
    startTime: string;
    winnerName: string;

    constructor(gameDB: GameDB) {
        this.startDate = gameDB.startDate;
        this.startTime = gameDB.startTime;
        this.winnerName = gameDB.winnerName;
    }

    isWinner(pseudonym: string): boolean {
        return pseudonym === this.winnerName;
    }
}

export class UserStats {
    userId: string;
    gamesPlayed: number;
    gamesWon: number;
    totalPoints: number;
    totalTimeMs: number;

    logins: Connection[];
    logouts: Connection[];
    games: Game[];

    constructor(userStatDB: UserStatsDB) {
        this.userId = userStatDB.userId;
        this.gamesPlayed = userStatDB.gamesPlayed;
        this.gamesWon = userStatDB.gamesWon;
        this.totalPoints = userStatDB.totalPoints;
        this.totalTimeMs = userStatDB.totalTimeMs;
        this.games = [];
        for (const game of userStatDB.games) {
            this.games.push(new Game(game));
        }
        this.logins = [];
        for (const login of userStatDB.logins) {
            this.logins.push(new Connection(login.date, login.time, login.isLogin));
        }
        this.logouts = [];
        for (const logout of userStatDB.logouts) {
            this.logouts.push(new Connection(logout.date, logout.time, logout.isLogin));
        }
    }

    getAveragePoints(): number {
        return Math.floor(this.totalPoints / this.gamesPlayed);
    }

    getAverageTime(): string {
        const averageTimeMs = this.totalTimeMs / this.gamesPlayed;

        const hour = Math.floor(averageTimeMs / (1000 * 60 * 60)) % 24;
        const minute = Math.floor(averageTimeMs / (1000 * 60)) % 60;
        const second = Math.floor(averageTimeMs / 1000) % 60;
        if (hour === 0) {
            const minutes = minute.toString() + 'm' + second.toString() + 's';
            return minutes;
        }
        const time = hour.toString() + 'h' + minute.toString() + 'm' + second.toString() + 's';
        return time;
    }
}
